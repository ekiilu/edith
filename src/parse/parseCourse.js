import React from "react"
import visit from "unist-util-visit"
import map from "unist-util-map"
import rehypeReact from "rehype-react"

import Link from "@material-ui/core/Link"
import Typography from "@material-ui/core/Typography"
import Table from "@material-ui/core/Table"
import TableHead from "@material-ui/core/TableHead"
import TableBody from "@material-ui/core/TableBody"
import TableRow from "@material-ui/core/TableRow"
import TableCell from "@material-ui/core/TableCell"

const renderAst = new rehypeReact({
  createElement: React.createElement,
  components: {
    h1: Typography,
    h2: Typography,
    h3: Typography,
    h4: Typography,
    h5: Typography,
    h6: Typography,
    a: Link,
    p: Typography,
    table: Table,
    thead: TableHead,
    tbody: TableBody,
    tr: TableRow,
    td: TableCell,
    th: TableCell,
  },
}).Compiler

export default ast => {
  let course = {
    elements: [],
  }

  visit(ast, "element", node => {
    if (node.tagName === "h1") {
      // Parse module title.
      let title = ""
      visit(node, "text", node => {
        title = title + node.value
      })
      course.title = title
    } else {
      const next = map(node, n => {
        const headers = ["h1", "h2", "h3", "h4", "h5", "h6"]
        if (headers.includes(n.tagName)) {
          return Object.assign({}, n, {
            tagName: headers[headers.indexOf(n.tagName)],
            properties: {
              ...n.properties,
              variant: headers[headers.indexOf(n.tagName)],
            },
          })
        } else if (n.tagName === "p") {
          return Object.assign({}, n, {
            properties: {
              ...n.properties,
              paragraph: true,
            },
          })
        }
        return n
      })
      course.elements.push(renderAst(next))
      return visit.SKIP
    }
  })

  return course
}
