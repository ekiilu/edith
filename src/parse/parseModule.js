import React from "react"
import visit from "unist-util-visit"
import map from "unist-util-map"
import rehypeReact from "rehype-react"

import Typography from "@material-ui/core/Typography"
import Table from "@material-ui/core/Table"
import TableHead from "@material-ui/core/TableHead"
import TableBody from "@material-ui/core/TableBody"
import TableRow from "@material-ui/core/TableRow"
import TableCell from "@material-ui/core/TableCell"
import Divider from "@material-ui/core/Divider"

import Link from "../components/Link"

const renderAst = new rehypeReact({
  createElement: React.createElement,
  components: {
    h1: Typography,
    h2: Typography,
    h3: Typography,
    h4: Typography,
    h5: Typography,
    h6: Typography,
    a: Link,
    p: Typography,
    table: Table,
    thead: TableHead,
    tbody: TableBody,
    tr: TableRow,
    td: TableCell,
    th: TableCell,
    hr: Divider,
  },
}).Compiler

export default ast => {
  let module = {}

  let units = []
  let currentUnit = {
    components: [],
  }
  visit(ast, "element", node => {
    if (node.tagName === "h1") {
      // Parse module title.
      let title = ""
      visit(node, "text", node => {
        title = title + node.value
      })
      module.title = title
    } else if (node.tagName === "h2") {
      // Close the previous unit.
      if (currentUnit.title) {
        units.push(currentUnit)
        currentUnit = {
          components: [],
        }
      }

      // Parse unit title.
      let title = ""
      visit(node, "text", node => {
        title = title + node.value
      })
      currentUnit.title = title
    } else {
      // Add to the current unit.
      if (currentUnit.title) {
        const next = map(node, n => {
          const headers = ["h1", "h2", "h3", "h4", "h5", "h6"]
          if (headers.includes(n.tagName)) {
            return Object.assign({}, n, {
              tagName: headers[headers.indexOf(n.tagName) - 1],
              properties: {
                ...n.properties,
                variant: headers[headers.indexOf(n.tagName) - 1],
              },
            })
          } else if (n.tagName === "p") {
            return Object.assign({}, n, {
              properties: {
                ...n.properties,
                paragraph: true,
              },
            })
          } else if (n.tagName === "th") {
            return Object.assign({}, n, {
              properties: {
                ...n.properties,
                component: "th",
              },
            })
          }
          return n
        })
        currentUnit.components.push(renderAst(next))
      }
      return visit.SKIP
    }
  })

  // Close last unit.
  if (currentUnit.title) {
    units.push(currentUnit)
  }

  module.units = units

  return module
}
