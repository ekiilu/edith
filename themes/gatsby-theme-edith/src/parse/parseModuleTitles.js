const visit = require("unist-util-visit")

/**
 * TODO: This duplicates a lot of what's in parseModule, but currently needed to be
 * able to be imported in gatsby-node.js
 */
exports.parseModuleTitles = ast => {
  let module = {}

  let units = []
  let currentUnit = {}
  visit(ast, "element", node => {
    if (node.tagName === "h1") {
      // Parse module title.
      let title = ""
      visit(node, "text", node => {
        title = title + node.value
      })
      module.title = title
    } else if (node.tagName === "h2") {
      // Close the previous unit.
      if (currentUnit.title) {
        units.push(currentUnit)
        currentUnit = {
          components: [],
        }
      }

      // Parse unit title.
      let title = ""
      visit(node, "text", node => {
        title = title + node.value
      })
      currentUnit.title = title
    } else {
      return visit.SKIP
    }
  })

  // Close last unit.
  if (currentUnit.title) {
    units.push(currentUnit)
  }

  module.units = units

  return module
}
