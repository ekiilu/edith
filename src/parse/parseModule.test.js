import u from "unist-builder"

import parse from "./parseModule"

describe("parse", () => {
  it("parses module titles", () => {
    const ast = u("root", [
      u("element", { tagName: "h1" }, [u("text", "Module title")]),
      u("text", "\n"),
      u("element", { tagName: "h2" }, [u("text", "Introduction")]),
      u("text", "\n"),
      u("element", { tagName: "p" }, [u("text", "This is about sheep.")]),
      u("text", "\n"),
      u("element", { tagName: "p" }, [
        u("text", "Lots of "),
        u("element", { tagName: "em" }, [u("text", "sheep")]),
      ]),
      u("text", "\n"),
      u("element", { tagName: "h2" }, [u("text", "Do this")]),
      u("text", "\n"),
      u("element", { tagName: "h2" }, [u("text", "Do that")]),
      u("text", "\n"),
      u("element", { tagName: "ul" }, [
        u("element", { tagName: "li" }, "First"),
        u("text", "\n"),
        u("element", { tagName: "li" }, "Second"),
        u("text", "\n"),
        u("element", { tagName: "li" }, "Third"),
      ]),
    ])

    expect(parse(ast)).toMatchSnapshot()
  })
})
