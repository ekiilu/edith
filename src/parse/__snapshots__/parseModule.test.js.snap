// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse parses module titles 1`] = `
Object {
  "title": "Module title",
  "units": Array [
    Object {
      "components": Array [
        <WithStyles(ForwardRef(Typography))
          paragraph={true}
        >
          This is about sheep.
        </WithStyles(ForwardRef(Typography))>,
        <WithStyles(ForwardRef(Typography))
          paragraph={true}
        >
          Lots of 
          <em>
            sheep
          </em>
        </WithStyles(ForwardRef(Typography))>,
      ],
      "title": "Introduction",
    },
    Object {
      "components": Array [],
      "title": "Do this",
    },
    Object {
      "components": Array [
        <ul>
          <li />
          

          <li />
          

          <li />
        </ul>,
      ],
      "title": "Do that",
    },
  ],
}
`;
